File: FxDSP/test/src/TestOptocoupler.m
//
//  TestOptocoupler.m
//  FxDSP
//
//  Created by Hamilton Kibbe on 4/5/14.
//  Copyright (c) 2014 Hamilton Kibbe. All rights reserved.
//

#import <XCTest/XCTest.h>
#include "Optocoupler.h"

@interface TestOptocoupler : XCTestCase

@end

@implementation TestOptocoupler


- (void)testStepResponse
{
    
    float step_function[100] = {0.0};
    float out[100] = {0.0};
    unsigned i;
    for (i = 25; i < 100; ++i)
    {
        step_function[i] = 1.0;
    }
    
    Opto* opto;
    opto = OptoInit(OPTO_LDR, 0.5, 2000);
    
    /* Process input */
    for (i = 0; i < 100; ++i)
    {
        out[i] = OptoTick(opto, step_function[i]);
    }
    OptoFree(opto);
}


- (void) testPulseResponse
{
    float pulse[200] = {0.0};
    float out[200] = {0.0};
    
    for (unsigned i = 50; i < 100; ++i)
    {
        pulse[i] = 1.0;
    }
    Opto* opto;
    opto = OptoInit(OPTO_PHOTOTRANSISTOR, 0.5, 1000);
    
    printf("\n\n\n\n\nPulseResp = [\n");
    /* Process input */
    for (unsigned i = 0; i < 200; ++i)
    {
        out[i] = OptoTick(opto, pulse[i]);
        printf("    %1.20f,\n", out[i]);
    }
    printf("];\n\n\n\n\n");
    OptoFree(opto);
}
@end

File: FxDSP/test/src/TestOptocoupler.cpp
//
//  TestOptocoupler.cpp
//  FxDSP
//
//  Created by Hamilton Kibbe on 5/3/15.
//  Copyright (c) 2015 Hamilton Kibbe. All rights reserved.
//

#include "Optocoupler.h"
#include <math.h>
#include <gtest/gtest.h>



TEST(OptocouplerSingle, Smoketest)
{
    const Opto_t types[2] = {OPTO_LDR, OPTO_PHOTOTRANSISTOR};
    float sinewave[10000];
    float out[10000];
    for (unsigned i = 0; i < 10000; ++i)
    {
        sinewave[i] = sinf((6000 * M_PI * i)/10000); // 3000Hz sinewave at 44100
    }
    for (unsigned i = 0; i < sizeof(types); ++i)
    {
        Opto* opto = OptoInit(types[i], 0.5, 44100);
        ASSERT_NE((void*)opto, (void*)NULL);

        OptoProcess(opto, out, sinewave, 10000);
        OptoFree(opto);
    }
}



TEST(OptocouplerDouble, Smoketest)
{
    const Opto_t types[2] = {OPTO_LDR, OPTO_PHOTOTRANSISTOR};
    double sinewave[10000];
    double out[10000];
    for (unsigned i = 0; i < 10000; ++i)
    {
        sinewave[i] = sin((6000 * M_PI * i)/10000); // 3000Hz sinewave at 44100
    }
    for (unsigned i = 0; i < sizeof(types); ++i)
    {
        OptoD* opto = OptoInitD(types[i], 0.5, 44100);
        ASSERT_NE((void*)opto, (void*)NULL);

        OptoProcessD(opto, out, sinewave, 10000);
        OptoFreeD(opto);
    }
}